src/node_jsrtapi.cc

```diff
@@ -170,6 +170,84 @@ void CHAKRA_CALLBACK JsObjectWrapWrapperBeforeCollectCallback(JsRef ref, void* c
+  do {                                                                  \
+    if (!(condition)) {                                                 \
+      return napi_set_last_error((status));                             \
+    }                                                                   \
+  } while(0)
+
+#define CHECK_ARG(arg)                                                  \
+  RETURN_STATUS_IF_FALSE((arg), napi_invalid_arg)
+
+#define CHECK_JSRT(expr)                                                \
+  do {                                                                  \
+    JsErrorCode err = (expr);                                           \
+    if (err != JsNoError) return napi_set_last_error(err);              \
+  } while(0)
+
+// This does not call napi_set_last_error because the expression
+// is assumed to be a NAPI function call that already did.
+#define CHECK_NAPI(expr)                                                \
+  do {                                                                  \
+    napi_status status = (expr);                                        \
+    if (status != napi_ok) return status;                               \
+  } while(0)
+
+// Static last error returned from napi_get_last_error_info
+napi_extended_error_info static_last_error;
+
+// Warning: Keep in-sync with napi_status enum
+const char* error_messages[] = {
+  nullptr,
+  "Invalid pointer passed as argument",
+  "An object was expected",
+  "A string was expected",
+  "A function was expected",
+  "Unknown failure",
+  "An exception is pending"
+};
+
+const napi_extended_error_info* napi_get_last_error_info() {
+  static_assert(sizeof(error_messages) / sizeof(*error_messages) == napi_status_last,
+    "Count of error messages must match count of error values");
+  assert(static_last_error.error_code < napi_status_last);
+
+  // Wait until someone requests the last error information to fetch the error message string
+  static_last_error.error_message = error_messages[static_last_error.error_code];
+
+  return &static_last_error;
+}
+
+napi_status napi_set_last_error(napi_status error_code,
+  uint32_t engine_error_code = 0,
+  void* engine_reserved = nullptr) {
+  static_last_error.error_code = error_code;
+  static_last_error.engine_error_code = engine_error_code;
+  static_last_error.engine_reserved = engine_reserved;
+
+  return error_code;
+}
+
+napi_status napi_set_last_error(JsErrorCode jsError, void* engine_reserved = nullptr) {
+  napi_status status;
+  switch (jsError) {
+    case JsNoError: status = napi_ok; break;
+    case JsErrorNullArgument:
+    case JsErrorInvalidArgument: status = napi_invalid_arg; break;
+    case JsErrorPropertyNotString: status = napi_string_expected; break;
+    case JsErrorArgumentNotObject: status = napi_object_expected; break;
+    case JsErrorScriptException:
+    case JsErrorInExceptionState: status = napi_pending_exception; break;
```

Comment:
Does this mean we don't need NAPI methods to check for pending exception before returning? Seems to.